llvm/test/CodeGen/AArch64/arm64-vhadd.ll

```diff
@@ -870,10 +866,6 @@ define <4 x i16> @hadd8_zext_lsr(<4 x i8> %src1, <4 x i8> %src2) {
 define <2 x i16> @hadd8x2_sext_asr(<2 x i8> %src1, <2 x i8> %src2) {
 ; CHECK-LABEL: hadd8x2_sext_asr:
 ; CHECK:       // %bb.0:
-; CHECK-NEXT:    shl.2s v1, v1, #24
-; CHECK-NEXT:    shl.2s v0, v0, #24
-; CHECK-NEXT:    sshr.2s v1, v1, #24
-; CHECK-NEXT:    sshr.2s v0, v0, #24
 ; CHECK-NEXT:    shadd.2s v0, v0, v1
```

Comment:
please can you confirm what's going on here?

Reply:
```
Type-legalized selection DAG: %bb.0 'hadd8x2_sext_asr:'
SelectionDAG has 13 nodes:
  t0: ch,glue = EntryToken
          t2: v2i32,ch = CopyFromReg t0, Register:v2i32 %0
        t21: v2i32 = sign_extend_inreg t2, ValueType:ch:v2i8
          t4: v2i32,ch = CopyFromReg t0, Register:v2i32 %1
        t22: v2i32 = sign_extend_inreg t4, ValueType:ch:v2i8
      t23: v2i32 = avgfloors t21, t22
    t24: v2i32 = sign_extend_inreg t23, ValueType:ch:v2i8
  t15: ch,glue = CopyToReg t0, Register:v2i32 $d0, t24
  t16: ch = AArch64ISD::RET_GLUE t15, Register:v2i32 $d0, t15:1



Combining: t24: v2i32 = sign_extend_inreg t23, ValueType:ch:v2i8
 ... into: t23: v2i32 = avgfloors t21, t22

Combining: t23: v2i32 = avgfloors t21, t22
Creating new node: t25: v2i32 = avgfloors t2, t4
 ... into: t25: v2i32 = avgfloors t2, t4

Combining: t25: v2i32 = avgfloors t2, t4

Combining: t16: ch = AArch64ISD::RET_GLUE t15, Register:v2i32 $d0, t15:1
Custom combining: skipping

Combining: t15: ch,glue = CopyToReg t0, Register:v2i32 $d0, t25

Combining: t14: v2i32 = Register $d0

Combining: t4: v2i32,ch = CopyFromReg t0, Register:v2i32 %1

Combining: t3: v2i32 = Register %1

Combining: t2: v2i32,ch = CopyFromReg t0, Register:v2i32 %0

Combining: t1: v2i32 = Register %0

Combining: t0: ch,glue = EntryToken

Optimized type-legalized selection DAG: %bb.0 'hadd8x2_sext_asr:'
SelectionDAG has 9 nodes:
  t0: ch,glue = EntryToken
      t2: v2i32,ch = CopyFromReg t0, Register:v2i32 %0
      t4: v2i32,ch = CopyFromReg t0, Register:v2i32 %1
    t25: v2i32 = avgfloors t2, t4
  t15: ch,glue = CopyToReg t0, Register:v2i32 $d0, t25
  t16: ch = AArch64ISD::RET_GLUE t15, Register:v2i32 $d0, t15:1
```

It appears that the issue is here. In the SelectionDAG graph, we observe that `t23` matches the rule we included in this pull request. However, instead of creating a `sext(avg(...))` node, it creates an `avg` node directly as `t25`. The cause of this current behavior remains unclear to me; perhaps more time is needed for investigation. Additionally, I have another question: which rule causes `t24` to combine into `t23`?

Reply:
```
Combining: t24: v2i32 = sign_extend_inreg t23, ValueType:ch:v2i8
 ... into: t23: v2i32 = avgfloors t21, t22
```
This will occur if value tracking has determined the known number of sign bits for the avgfloors result already matches/exceeds what the sign_extend_inreg node would do.

Reply:
Found the problem - m_SExt is matching ISD::SIGN_EXTEND_INREG nodes, which isn't what we want here. 

@mshockwave I think this is a bug as we can't easily discriminate between ISD::SIGN_EXTEND and ISD::SIGN_EXTEND_INREG.

Reply:
I believe SelectionDAGPatternMatchTest.cpp will need updating as well - are you OK with creating a new PR to handle this first?